### PR TITLE
PhysicalDiskInfo class #455

### PR DESCRIPTION
**Pull request by Yomodo**
Hi @alphaleonis, I'd appreciate a review on this one, for a v2.3.
I think it's ready to go.

### CODE REVIEW

File: AlphaFS/Device/DriveInfo Class/DriveInfo.GetDeviceInfo.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+{
+   public sealed partial class DriveInfo
+   {
+      /// <summary>Retrieves information about the file system and volume associated with the specified root file or directorystream.</summary>
+      [SuppressMessage("Microsoft.Maintainability", "CA1502:AvoidExcessiveComplexity")]
+      [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+      [SecurityCritical]
+      private object GetDeviceInfo(int type, int mode)
+      {
+         //try
```
Comment (alphaleonis) [inline AlphaFS/Device/DriveInfo Class/DriveInfo.GetDeviceInfo.cs:36]: Remove commented try/catch or use it.

File: AlphaFS/Device/DriveInfo Class/DriveInfo.GetDeviceInfo.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+namespace Alphaleonis.Win32.Filesystem
+{
+   public sealed partial class DriveInfo
+   {
+      /// <summary>Retrieves information about the file system and volume associated with the specified root file or directorystream.</summary>
+      [SuppressMessage("Microsoft.Maintainability", "CA1502:AvoidExcessiveComplexity")]
+      [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+      [SecurityCritical]
+      private object GetDeviceInfo(int type, int mode)
```
Comment (alphaleonis) [inline AlphaFS/Device/DriveInfo Class/DriveInfo.GetDeviceInfo.cs:34]: Can we change type/mode to private enums to make it clear what this method does, it is very confusing with the integers. Or at the very least, document it.

**Review by alphaleonis — COMMENTED** [review]
There is really no way I can do a review of this, with 332 changed files and over 250+ commits, and no assicated issues or any hint as to what is really done here? 

Seems that most classes have been split into files per method, which is fine, but makes it impossible to track down what has changed.

I know that sometime major organizational changes like this is needed, but it should be kept in separate pull-requests/branches from any feature additions so that it is possible to actually review and follow what has been done.

Can you give me some description of the changes that are actually made and what, if anything, you want me to look closer at?

**Comment by Yomodo** [comment]
Hm yeah, perhaps a bit too much for a simple "review", and perhaps not the best word for it.

The majority of the commits are all about the Device.PhysicalDiskInfo class and related members,
a new addition to AlphaFS, and not so much about radically changing the existing base code, other than splitting files/methods.

So the review is not about comparing existing base code that has been improved,
but more of an approval for this new AlphaFS member.

PhysicalDiskInfo role is to kind of act like a DiskInfo, but aimed at, well, physical disks.
The issues you pointed out is exactly what I mean.

**Comment by alphaleonis** [comment]
Alright, I'll have another look at the `PhysicalDiskInfo` class shortly then and get back to you. :)

File: AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      private Collection<int> _partitionIndexCollection;
+      private Collection<string> _volumeGuidCollection;
+      private Collection<string> _logicalDriveCollection;
+
+      #endregion // Fields
+
+
+      #region Constructors
+
+      private PhysicalDiskInfo()
```
Comment (alphaleonis) [inline AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs:46]: Constructor appears unused. Remove.

File: AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      {
+      }
+
+
+      /// <summary>[AlphaFS] Initializes an instance from a physical disk device number.</summary>
+      /// <param name="deviceNumber">A device number that indicates a physical disk on the Computer.</param>
```
Comment (alphaleonis) [inline AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs:52]: Can we add a bit more description on what a "device number" actually is`?  Is it like the comment on `DISK_EXTENT.DiskNumber`?

File: AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+         CreatePhysicalDiskInfoInstance(deviceNumber, null, storageDeviceInfo, deviceInfo);
+      }
+
+      #endregion // Constructors
+
+
+      #region Properties
+
+      /// <summary>An initialized <see cref="Filesystem.DeviceInfo"/> instance.</summary>
+      private DeviceInfo DeviceInfo { get; set; }
```
Comment (alphaleonis) [inline AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs:91]: Shouldn't this be `private set` ?

File: AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+
+      /// <summary>The device description.</summary>
+      public string DeviceDescription
+      {
+         get { return null != DeviceInfo ? DeviceInfo.DeviceDescription : string.Empty; }
```
Comment (alphaleonis) [inline AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs:97]: These constructs could be abbreviated to eg. `DeviceInfo?.DeviceDescription ?? String.Empty`

File: AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.ContainsVolume.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+using System;
+using System.Linq;
+using Alphaleonis.Win32.Filesystem;
+
+namespace Alphaleonis.Win32.Device
+{
+   public sealed partial class PhysicalDiskInfo
+   {
+      /// <summary>Checks if the volume/logical drive is located on the physical disk.
+      /// <para>A drive path such as: <c>C</c>, <c>C:</c> or <c>C:\</c></para>
```
Comment (alphaleonis) [inline AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.ContainsVolume.cs:31]: The two paragraphs here seems to belong with the description of the `devicePath` parameter (where the also appear to be. Remove them from the summary.

File: AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+      /// <summary>The storage device information.</summary>
+      public StorageDeviceInfo StorageDeviceInfo { get; private set; }
+
+
+      /// <summary>The storage device partition information.</summary>
+      public StoragePartitionInfo StoragePartitionInfo { get; private set; }
+
+
+      /// <summary>An <see cref="IEnumerable{String}"/> of volume <see cref="Guid"/> strings that are located on the physical disk.</summary>
+      public IEnumerable<string> VolumeGuids { get; private set; }
```
Comment (alphaleonis) [inline AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs:155]: What is a volume Guid?  If it's a Guid, then why is the property `IEnumerable<string>` rather than `IEnumerable<Guid>`?

File: AlphaFS/Device/Storage/Device.StorageAdapterInfo.cs [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+      /// <summary>The device number connected to this storage adapter, starting at 0.</summary>
+      public int DeviceNumber { get; private set; }
+
+
+      /// <summary>Specifies the maximum number of bytes the storage adapter can transfer in a single operation.</summary>
+      public int MaximumTransferBytes { get; private set; }
+
+
+#if DEBUG
+      /// <summary>Specifies the maximum number of bytes the storage adapter can transfer in a single operation, formatted as a unit size.</summary>
```
Comment (alphaleonis) [inline AlphaFS/Device/Storage/Device.StorageAdapterInfo.cs:107]: Why is this property enclosed in `#if DEBUG` tags? Yet `StorageDeviceInfo.TotalSizeUnitSize` is not?

File: AlphaFS/Device/Storage/Device.StorageDeviceInfo.cs [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+      #endregion // Properties
+
+
+      #region Methods
+
+      /// <summary>Returns storage device as: "DeviceType DeviceNumber:PartitionNumber VendorId ProductId".</summary>
+      /// <returns>Returns a string that represents this instance.</returns>
+      public override string ToString()
+      {
+         return string.Format(CultureInfo.CurrentCulture, "Device: {0}:{1} {2}", DeviceNumber.ToString(CultureInfo.InvariantCulture), PartitionNumber.ToString(CultureInfo.InvariantCulture), (VendorId + " " + ProductId + " " + ProductRevision).Trim()).Trim();
```
Comment (alphaleonis) [inline AlphaFS/Device/Storage/Device.StorageDeviceInfo.cs:138]: Why do we format with CurrentCulture, and then every number you pass in is formatted with InvariantCulture explicitly?  Prefer string interpolation in these cases (there is nothing culture sensitive in this particular string, so would probably be clearer as `return $"Device: {DeviceNumber}:{PartitionNumber} {VendorId} {ProductId} {ProductRevision}".Trim();`

File: AlphaFS/Device/Storage/Device.StoragePartitionInfo.cs [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+      private ulong _gptStartingUsableOffset;
+      private ulong _gptUsableLength;
+
+      #endregion // Private Fields
+
+
+      #region Constructors
+
+      /// <summary>[AlphaFS] Initializes an empty StoragePartitionInfo instance.</summary>
+      public StoragePartitionInfo()
```
Comment (alphaleonis) [inline AlphaFS/Device/Storage/Device.StoragePartitionInfo.cs:48]: Why do we need a public default constructor? None of the properties are settable anyway?

File: AlphaFS/Device/Storage/Device.StorageDeviceInfo.cs [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+
+
+      #endregion // Private Fields
+
+
+      #region Constructors
+
+      /// <summary>[AlphaFS] Initializes an empty StorageDeviceInfo instance.</summary>
+      public StorageDeviceInfo()
+      {
```
Comment (alphaleonis) [inline AlphaFS/Device/Storage/Device.StorageDeviceInfo.cs:44]: Why do we need a public default constructor? None of the properties are settable anyway?

File: AlphaFS/Device/Storage/Device.StorageAdapterInfo.cs [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+   /// <summary>[AlphaFS] Provides access to adapter information of a storage device.</summary>
+   [Serializable]
+   [SecurityCritical]
+   public sealed class StorageAdapterInfo
+   {
+      #region Constructors
+
+      /// <summary>[AlphaFS] Initializes an empty StorageAdapterInfo instance.</summary>
+      public StorageAdapterInfo()
+      {
```
Comment (alphaleonis) [inline AlphaFS/Device/Storage/Device.StorageAdapterInfo.cs:38]: Why do we need a public default constructor? None of the properties are settable anyway?

File: AlphaFS/Device/Storage/Device.StorageAdapterInfo.cs [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+         MaximumTransferBytes = (int) adapter.MaximumTransferLength;
+      }
+
+      #endregion // Constructors
+
+
+      #region Properties
+
+      /// <summary>When <c>true</c>, the storage adapter supports synchronous transfers as a way of speeding up I/O.</summary>
+      public bool AcceleratedTransfer { get; private set; }
```
Comment (alphaleonis) [inline AlphaFS/Device/Storage/Device.StorageAdapterInfo.cs:70]: On all properties in pretty much all classes, remove `private set` if the property is read-only, and only set from the ctor.

File: AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+using System.Collections.ObjectModel;
+using System.Security;
+using Alphaleonis.Win32.Filesystem;
+
+namespace Alphaleonis.Win32.Device
+{
+   /// <summary>[AlphaFS] Provides access to information of a physical disk on the Computer.</summary>
+   [Serializable]
+   [SecurityCritical]
+   public sealed partial class PhysicalDiskInfo
```
Comment (alphaleonis) [inline AlphaFS/Device/PhysicalDisk Class/PhysicalDiskInfo.cs:33]: Can't find any updates to the CHANGELOG?  I'm assuming this is a new public class?

File: AlphaFS/Device/Local Class/Device.Local.cs [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN 
+ *  THE SOFTWARE. 
+ */
+
+namespace Alphaleonis.Win32.Device
+{
+   /// <summary>[AlphaFS] Provides static methods to retrieve device resource information from the local host.</summary>
+   public static partial class Local
```
Comment (alphaleonis) [inline AlphaFS/Device/Local Class/Device.Local.cs:25]: The name `Local` is way to generic, and doesn't really say anything about what the class does. A name like `Device` would probably be better?  Especially since it apparently can retrieve things from a remote host as well?

File: AlphaFS/Device/Local Class/Local.EnumerateDevices.cs [hunk 17]
```diff
@@ -0,0 +1,344 @@
...
+      }
+
+
+      /// <summary>[AlphaFS] Enumerates all available devices of type <see cref="DeviceGuid"/> on the local host.</summary>
+      /// <returns>Returns an <see cref="IEnumerable{DeviceInfo}"/> collection from the local host.</returns>
+      /// <param name="deviceGuid">One or more <see cref="DeviceGuid"/> device guids.</param>
+      [SecurityCritical]
+      public static IEnumerable<DeviceInfo> EnumerateDevices(DeviceGuid[] deviceGuid)
+      {
+         return EnumerateDevicesCore(null, deviceGuid, true);
```
Comment (alphaleonis) [inline AlphaFS/Device/Local Class/Local.EnumerateDevices.cs:51]: Could we make the input array a `params` array instead and hence remove the previous method?

File: AlphaFS/Device/Local Class/Local.EnumerateDevices.cs [hunk 18]
```diff
@@ -0,0 +1,344 @@
...
+      /// <returns>Returns an <see cref="IEnumerable{DeviceInfo}"/> collection for the specified <paramref name="hostName"/>.</returns>
+      /// <remarks>
+      ///   MSDN Note: Beginning in Windows 8 and Windows Server 2012 functionality to access remote machines has been removed.
+      ///   You cannot access remote machines when running on these versions of Windows.
+      ///   <para>http://msdn.microsoft.com/en-us/library/windows/hardware/ff537948%28v=vs.85%29.aspx</para>
+      /// </remarks>
+      /// <param name="hostName">The name of the local or remote host on which the device resides. <c>null</c> refers to the local host.</param>
+      /// <param name="deviceGuid">One or more <see cref="DeviceGuid"/> device guids.</param>
+      [SecurityCritical]
+      public static IEnumerable<DeviceInfo> EnumerateDevices(string hostName, DeviceGuid[] deviceGuid)
```
Comment (alphaleonis) [inline AlphaFS/Device/Local Class/Local.EnumerateDevices.cs:81]: Same here, `params` array?

File: AlphaFS/Device/Local Class/Local.EnumerateDevices.cs [hunk 19]
```diff
@@ -0,0 +1,344 @@
...
+      /// <summary>[AlphaFS] Enumerates all available devices of type <see cref="DeviceGuid"/> on the local or remote host.</summary>
+      /// <returns>Returns an <see cref="IEnumerable{DeviceInfo}"/> collection for the specified <paramref name="hostName"/>.</returns>
+      /// <remarks>
+      ///   MSDN Note: Beginning in Windows 8 and Windows Server 2012 functionality to access remote machines has been removed.
+      ///   You cannot access remote machines when running on these versions of Windows.
+      ///   <para>http://msdn.microsoft.com/en-us/library/windows/hardware/ff537948%28v=vs.85%29.aspx</para>
+      /// </remarks>
+      /// <param name="hostName">The name of the local or remote host on which the device resides. <c>null</c> refers to the local host.</param>
+      /// <param name="deviceGuid">One or more <see cref="DeviceGuid"/> device guids.</param>
+      /// <param name="getAllProperties"><c>true</c> to retrieve all device properties.</param>
```
Comment (alphaleonis) [inline AlphaFS/Device/Local Class/Local.EnumerateDevices.cs:96]: What does `getAllProperties` mean? What is retrieved otherwise?

**Review by alphaleonis — CHANGES_REQUESTED** [review]
Don't forget to update CHANGELOG.

Added a bunch of comments to various things that caught my attention. But it is really, really difficult to review a changeset of this size.
In the future, please keep refactorings (such as splitting stuff into many new files) as separate commits/pull requests from any actual changes to code, and please request code-reviews more often on smaller changes, rather than something as big as this, where it's impossible to get an overview of what has actually changed in the public surface area.

In short though the PhysicalDiskInfo seems like a good adition, though the name of the associated class `Local` must be changed to something more descriptive.
